packages/block-editor/src/components/block-lock/toolbar.js

```diff
@@ -33,10 +34,23 @@ export default function BlockLockToolbar( { clientId, wrapperRef } ) {
 	useEffect( () => {
 		if ( isFirstRender.current ) {
 			isFirstRender.current = false;
+			hasModalOpened.current = false;
 			return;
 		}
 
-		if ( ! isModalOpen && shouldHideBlockLockUI ) {
+		if ( isModalOpen && ! hasModalOpened.current ) {
+			hasModalOpened.current = true;
+		}
```

Comment:
(Optional) Putting together a couple of suggestions left above, we could also split the logic into three smaller `useEffect`s — one for the "first render" logic, one for the "has modal ever opened" logic, and one for focus restoration

```diff
diff --git a/packages/block-editor/src/components/block-lock/toolbar.js b/packages/block-editor/src/components/block-lock/toolbar.js
index 3385dd1b15..2c69fe3143 100644
--- a/packages/block-editor/src/components/block-lock/toolbar.js
+++ b/packages/block-editor/src/components/block-lock/toolbar.js
@@ -28,42 +28,47 @@ export default function BlockLockToolbar( { clientId, wrapperRef } ) {
 	const shouldHideBlockLockUI =
 		! canLock || ( canEdit && canMove && canRemove );
 
-	// Restore focus manually on the first focusable element in the toolbar
-	// when the block lock modal is closed and the block is not locked anymore.
-	// See https://github.com/WordPress/gutenberg/issues/51447
 	useEffect( () => {
 		if ( isFirstRender.current ) {
 			isFirstRender.current = false;
-			hasModalOpened.current = false;
-			return;
 		}
+	}, [] );
 
+	useEffect( () => {
 		if ( isModalOpen && ! hasModalOpened.current ) {
 			hasModalOpened.current = true;
 		}
+	}, [ isModalOpen ] );
 
+	// Restore focus manually on the first focusable element in the toolbar
+	// when the block lock modal is closed and the block is not locked anymore.
+	// See https://github.com/WordPress/gutenberg/issues/51447
+	useEffect( () => {
 		// We only want to allow this effeect to happen if the modal has been opened.
 		// The issue is when we're returning focus from the block lock modal to a toolbar,
 		// so it can only happen after a modal has been opened. Without this, the toolbar
 		// will steal focus on rerenders.
 		if (
-			hasModalOpened.current &&
-			! isModalOpen &&
-			shouldHideBlockLockUI
+			isFirstRender.current ||
+			! hasModalOpened.current ||
+			isModalOpen ||
+			! shouldHideBlockLockUI
 		) {
-			focus.focusable
-				.find( wrapperRef.current, {
-					sequential: false,
-				} )
-				.find(
-					( element ) =>
-						element.tagName === 'BUTTON' &&
-						element !== lockButtonRef.current
-				)
-				?.focus();
+			return;
 		}
+
+		focus.focusable
+			.find( wrapperRef.current, {
+				sequential: false,
+			} )
+			.find(
+				( element ) =>
+					element.tagName === 'BUTTON' &&
+					element !== lockButtonRef.current
+			)
+			?.focus();
 		// wrapperRef is a reference object and should be stable
-	}, [ isModalOpen, shouldHideBlockLockUI, hasModalOpened, wrapperRef ] );
+	}, [ isModalOpen, shouldHideBlockLockUI, wrapperRef ] );
 
 	if ( shouldHideBlockLockUI ) {
 		return null;

```